README

                                      JSch

                                Java Secure Channel
                         by ymnk@jcraft.com, JCraft,Inc.

                            http://www.jcraft.com/jsch/

Last modified: Thu Mar 18 13:58:16 UTC 2015


Description
===========
JSch is a pure Java implementation of SSH2.  JSch allows you to
connect to an sshd server and use port forwarding, X11 forwarding,
file transfer, etc., and you can integrate its functionality
into your own Java programs. JSch is licensed under BSD style license.


Documentation
=============
* README files all over the source tree have info related to the stuff
  in the directories.
* ChangeLog: what changed from the previous version?


Directories & Files in the Source Tree
======================================
* src/com/ has source trees of JSch
* example/ has some samples, which demonstrate the usages.
* tools/ has scripts for Ant.


Why JSch?
==========
Our intension in developing this stuff is to enable users of our pure
java X servers, WiredX(http://wiredx.net/) and WeirdX, to enjoy secure X
sessions.  Our efforts have mostly targeted the SSH2 protocol in relation
to X Window System and X11 forwarding.  Of course, we are also interested in
adding other functionality - port forward, file transfer, terminal emulation, etc.


Features
========
* JSch is in pure Java, but it depends on JavaTM Cryptography
  Extension (JCE).  JSch is know to work with:
  o J2SE 1.4.0 or later (no additional libraries required).
  o J2SE 1.3 and Sun's JCE reference implementation that can be
    obtained at http://java.sun.com/products/jce/
  o J2SE 1.2.2 and later and Bouncycastle's JCE implementation that
    can be obtained at http://www.bouncycastle.org/
* SSH2 protocol support.
* Key exchange: diffie-hellman-group-exchange-sha1,
                diffie-hellman-group1-sha1,
                diffie-hellman-group14-sha1,
                diffie-hellman-group-exchange-sha256,
                ecdh-sha2-nistp256,
                ecdh-sha2-nistp384,
                ecdh-sha2-nistp521
* Cipher: blowfish-cbc,3des-cbc,aes128-cbc,aes192-cbc,aes256-cbc
          3des-ctr,aes128-ctr,aes192-ctr,aes256-ctc,
          arcfour,arcfour128,arcfour256
* MAC: hmac-md5,hmac-md5-96,hmac-sha1,hmac-sha1-96
* Host key type: ssh-dss,ssh-rsa,
                 ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521
* Userauth: password
* Userauth: publickey(DSA,RSA,ECDSA)
* Userauth: keyboard-interactive
* Userauth: gssapi-with-mic
* X11 forwarding.
* xauth spoofing.
* connection through HTTP proxy.
* connection through SOCKS5, SOCKS4 proxy.
* port forwarding.
* stream forwarding.
* signal sending.
  The unofficial patch for sshd of openssh will be find in the thread
  http://marc.theaimsgroup.com/?l=openssh-unix-dev&m=104295745607575&w=2
* envrironment variable passing.
* remote exec.
* generating DSA and RSA key pairs.
* supporting private keys in OpenSSL(traditional SSLeay) and PKCS#8 format.
* SSH File Transfer Protocol(version 0, 1, 2, 3)
* partial authentication
* packet compression: zlib, zlib@openssh.com
  JZlib(http://www.jcraft.com/jzlib/) has been used.
* hashed known_hosts file.
* NONE Cipher switching.
  http://www.psc.edu/networking/projects/hpn-ssh/none.php
* JSch is licensed under BSD style license(refer to LICENSE.txt).


How To Try
==========
This archive does not include java byte code, so please compile
the source code by your self.
  $ cd jsch-?.?.?/src
  $ javac com/jcraft/jsch/*java com/jcraft/jsch/jce/*java com/jcraft/jzlib/*.java
'/examples/' directory has included some samples to demonstrate what
JSch can do.  Please refer to '/examples/README' file.


AES ciphering
==========
JSch supports aes128-cbc,aes192-cbc,aes256-cbc,aes128-ctr,aes192-ctr,
aes256-ctr but you require AES support in your J2SE to choose some of them.
If you are using Sun's J2SE, J2SE 1.4.2 or later is required.
And then, J2SE 1.4.2(or later) does not support aes256 by the default,
because of 'import control restrictions of some countries'.
We have confirmed that by applying
  "Java Cryptography Extension (JCE)
  Unlimited Strength Jurisdiction Policy Files 1.4.2"
on
  http://java.sun.com/j2se/1.4.2/download.html#docs
we can enjoy 'aes256-cbc,aes256-ctr'.


Stream Forwarding
=================
JSch has a unique functionality, Stream Forwarding.
Stream Forwarding allows you to plug Java I/O streams directly into a remote TCP
port without assigning and opening a local TCP port.
In port forwarding, as with the -L option of ssh command, you have to assign
and open a local TCP port and that port is also accessible by crackers
on localhost.  In some case, that local TCP port may be plugged to a
secret port via SSH session.
A sample program, /example/StreamForwarding.java , demonstrates
this functionality.


Generating Authentication Keys
==============================
JSch allows you to generate DSA and RSA key pairs, which are in OpenSSH format.
Please refer to 'examples/KeyGen.java'.


Packet Compression
==================
According to the draft from IETF sesch working group, the packet
compression can be applied to each data stream directions; from sshd
server to ssh client and from ssh client to sshd server.  So, jsch
allows you to choose which data stream direction will be compressed or not.
For example, in X11 forwarding session, the packet compression for data
stream from sshd to ssh client will save the network traffic, but
usually the traffic from ssh client to sshd is light, so by omitting
the compression for this direction, you may be able to save some CPU time.
Please refer to a sample program 'examples/Compression.java'.


Property
========
By setting properties, you can control the behavior of jsch.
Here is an example of enabling the packet compression,

      Session session=jsch.getSession(user, host, 22);
      java.util.Properties config=new java.util.Properties();
      config.put("compression.s2c", "zlib,none");
      config.put("compression.c2s", "zlib,none");
      session.setConfig(config);
      session.connect();

Current release has supported following properties,
* compression.s2c: zlib, none
  default: none
  Specifies whether to use compression for the data stream
  from sshd to jsch.  If "zlib,none" is given and the remote sshd does
  not allow the packet compression, compression will not be done.
* compression.c2s: zlib, none
  default: none
  Specifies whether to use compression for the data stream
  from jsch to sshd.
* StrictHostKeyChecking: ask | yes | no
  default: ask
  If this property is set to ``yes'', jsch will never automatically add
  host keys to the $HOME/.ssh/known_hosts file, and refuses to connect
  to hosts whose host key has changed.  This property forces the user
  to manually add all new hosts.  If this property is set to ``no'',
  jsch will automatically add new host keys to the user known hosts
  files.  If this property is set to ``ask'', new  host keys will be
  added to the user known host files only after the user has confirmed
  that is what they really want to do, and jsch will refuse to connect
  to hosts whose host key has changed.


TODO
====
* re-implementation with java.nio.
* replacing ciphering, hash by JCE with pure Java code.
* SSH File Transfer Protocol version 4.
* error handling.


Copyrights & Disclaimers
========================
JSch is copyrighted by ymnk, JCraft,Inc. and is licensed through BSD style license.
Read the LICENSE.txt file for the complete license.


Credits and Acknowledgments
============================
JSch has been developed by ymnk@jcraft.com and it can not be hacked
without several help.
* First of all, we want to thank JCE team at Sun Microsystems.
  For long time, we had planed to implement SSH2 in pure Java,
  but we had hesitated to do because tons of work must be done for
  implementing ciphers, hashes, etc., from the scratch.
  Thanks to newly added functionalities to J2SE 1.4.0, we could
  start this project.
* We appreciate the OpenSSH project.
  The options '-ddd' of sshd, '---vvv' of ssh and the compile options
  '-DPACKET_DEBUG', '-DDEBUG_KEXDH' and  '-DDEBUG_KEX' were very
  useful in debugging JSch.
* We appreciate IETF sesch working group and SSH Communications Security Corp.
  Without the standardization of the protocol, we could not get the
  chance to implement JSch.
* We appreciate Seigo Haruyama(http://www.unixuser.org/~haruyama/),
  who are interpreting drafts of SSH2 protocol in Japanese.
  His works were very useful for us to understand the technical terms
  in our native language.
* We also appreciate SourceForge.net's awesome service to the
  Open Source Community.


If you have any comments, suggestions and questions, write us
at jsch@jcraft.com


``SSH is a registered trademark and Secure Shell is a trademark of
SSH Communications Security Corp (www.ssh.com)''.


ChangeLog
ChangeLog of JSch
====================================================================
Last modified: Tue Aug 30 06:47:52 UTC 2016


Changes since version 0.1.53:
- bugfix: fixed CVS-2016-5725
          Refer to following links,
            http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2016-5725
            https://github.com/tintinweb/pub/tree/master/pocs/cve-2016-5725
          Thanks a lot for tintinweb's contributions.
- bugfix: sftp-put may send the garbage data in some rare case.
- bugfix: fixed a deadlock bug in KnownHosts#getHostKey().
- bugfix: SftpProgressMonitor#init() was not invoked in sftp-put
          by using the output-stream.
- change: KnownHosts#setKnownHosts() should accept the non-existing file.
- change: excluding the user interaction time from the timeout value.
- change: addressing SFTP slow file transfer speed with Titan FTP.
- change: updating copyright messages; 2015 -> 2016


Changes since version 0.1.52:
- bugfix: the rekey initiated by the remote may crash the session.
- change: Logjam: use ecdh-sha2-nistp* if available,
          ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,
          diffie-hellman-group14-sha1,
          diffie-hellman-group-exchange-sha256,
          diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1
- change: Logjam: diffie-hellman-group-exchange-sha256 and
          diffie-hellman-group-exchange-sha1 will use 2048-bit key on
          Java8's SunJCE, thanks to JDK-6521495 and JDK-7044060.
- change: key words for OpenSSH's config file should be case-insensitive.
- change: there should be the host name in "WARNING: REMOTE HOST
          IDENTIFICATION HAS CHANGED" message.


Changes since version 0.1.51:
- bugfix: resource leak: duplicate keys in LocalIdentityRepository.
- feature: added the support for SSH ECC defined in RFC5656,
           ecdsa-sha2-nistp256,
           ecdsa-sha2-nistp384,
           ecdsa-sha2-nistp521,
           ecdh-sha2-nistp256,
           ecdh-sha2-nistp384,
           ecdh-sha2-nistp521
           This functionality requires Java7 or later.
- feature: added the support for server host keys in
           ecdsa-sha2-nistp256,
           ecdsa-sha2-nistp384,
           ecdsa-sha2-nistp521
- feature: generating key-pairs in
           ecdh-sha2-nistp256,
           ecdh-sha2-nistp384,
           ecdh-sha2-nistp521
- change: aes192-ctr, aes256-ctr and
          diffie-hellman-group-exchange-sha256 have been enabled
          by the default.
- change: key exchange methods, ecdh-sha2-nistp256,
          ecdh-sha2-nistp384 and ecdh-sha2-nistp521 have been enabled
          by the default.
- change: the support for host keys in ecdsa-sha2-nistp256,
          ecdsa-sha2-nistp384 and ecdsa-sha2-nistp521 have been enabled
          by the default.
- change: 'examples/KeyGen.java' demonstrates how to generate
          ecdsa-sha2-* key-pairs.
- change: updating copyright messages; 2014 -> 2015
- TODO: The ECC support is not functional on Java6 with BouncyCastle.


Changes since version 0.1.50:
- bugfix: reproducibility of "verify: false".   FIXED.
          Hundreds of thousands of connections had caused that exception.
- bugfix: resource leaks at the failure of making local port forwarding. FIXED.
- bugfix: NPE in connecting to the non-standard TCP port.  FIXED.
          This problem had appeared if a host-key does not exist in
	  "known_host" file.
- bugfix: TCP connection may not be dropped if error messages from
	  the remote are too long.                          FIXED.
- bugfix: SftpATTRS#getAtimeString() returns the wrong string. FIXED.
- bugfix: bytes to be added by SSH_MSG_CHANNEL_WINDOW_ADJUST must be in
	  unsigned integer. FIXED.
- bugfix: Util#checkTilde() should not convert a tilde in
          "C:\PROGRA~1\". FIXED.
- bugfix: A long long command for ChannelExec will cause
          an ArrayIndexOutOfBoundsException. FIXED.
- bugfix: ChannelSftp should not send bulk request greedily even if the remote
          window has the enough space.  FIXED.
- bugfix: Util.createSocket() should throw an exception with 'cause'. FIXED.
- bugfix: failed to parse .ssh/config in the EBCDIC environment. FIXED.
- bugfix: com.jcraft.jsch.jcraft.HMACSHA1(used only for MacOSX) is not
          reusable.  FIXED.
- bugfix: NPE caused by the delayed response for channel opening
          requests. FIXED.
- bugfix: hung-up in uploading huge data to ProFTPd without the config
          'SFTPClientMatch "JSCH.*" channelWindowSize 1GB'  FIXED.
- bugfix: Cipher#init() may cause an infinite loop with 100% cpu use due to
          https://bugs.openjdk.java.net/browse/JDK-8028627  FIXED.
- bugfix: in some case, JSche#setKnowHosts(InputStream stream) may fail
          to close the given stream.  FIXED
- change: com.jcraft.jsch.jcraft.HMAC* will not be used.
          It seems Java6 on Mac OS X has fixed some memory leak bug in JCE,
	  so there is no reason to use c.j.j.j.HMAC* introduced at 0.1.30.
- change: updating copyright messages; 2013 -> 2014
- change: allowed to create the symbolic/hard link to the relative path by
          ChannelSftp#symlink(String oldpath, String newpath)
          ChannelSftp#hardlink(String oldpath, String newpath)

src/main/java/com/jcraft/jsch/JSch.java
/* -*-mode:java; c-basic-offset:2; indent-tabs-mode:nil -*- */
/*
Copyright (c) 2002-2016 ymnk, JCraft,Inc. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  1. Redistributions of source code must retain the above copyright notice,
     this list of conditions and the following disclaimer.

  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the distribution.

  3. The names of the authors may not be used to endorse or promote products
     derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL JCRAFT,
INC. OR ANY CONTRIBUTORS TO THIS SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

package com.jcraft.jsch;

import java.io.InputStream;
import java.util.Hashtable;
import java.util.Vector;

public class JSch {
    /**
     * The version number.
     */
    public static final String VERSION = "0.1.54";

    private static final Hashtable<String, String> config = new Hashtable<>();

    static {
        config.put("kex", "ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1");
        config.put("server_host_key", "ssh-rsa,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521");
        config.put("cipher.s2c", "aes128-ctr,aes128-cbc,3des-ctr,3des-cbc,blowfish-cbc,aes192-ctr,aes192-cbc,aes256-ctr,aes256-cbc");
        config.put("cipher.c2s", "aes128-ctr,aes128-cbc,3des-ctr,3des-cbc,blowfish-cbc,aes192-ctr,aes192-cbc,aes256-ctr,aes256-cbc");

        config.put("mac.s2c", "hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha1-96,hmac-md5-96");
        config.put("mac.c2s", "hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha1-96,hmac-md5-96");
        config.put("compression.s2c", "none");
        config.put("compression.c2s", "none");

        config.put("lang.s2c", "");
        config.put("lang.c2s", "");

        config.put("compression_level", "6");

        config.put("diffie-hellman-group-exchange-sha1", "com.jcraft.jsch.DHGEX");
        config.put("diffie-hellman-group1-sha1", "com.jcraft.jsch.DHG1");
        config.put("diffie-hellman-group14-sha1", "com.jcraft.jsch.DHG14");    // available since JDK8.
        config.put("diffie-hellman-group-exchange-sha256", "com.jcraft.jsch.DHGEX256"); // available since JDK1.4.2.
        // On JDK8, 2048bits will be used.
        config.put("ecdsa-sha2-nistp256", "com.jcraft.jsch.jce.SignatureECDSA");
        config.put("ecdsa-sha2-nistp384", "com.jcraft.jsch.jce.SignatureECDSA");
        config.put("ecdsa-sha2-nistp521", "com.jcraft.jsch.jce.SignatureECDSA");

        config.put("ecdh-sha2-nistp256", "com.jcraft.jsch.DHEC256");
        config.put("ecdh-sha2-nistp384", "com.jcraft.jsch.DHEC384");
        config.put("ecdh-sha2-nistp521", "com.jcraft.jsch.DHEC521");

        config.put("ecdh-sha2-nistp", "com.jcraft.jsch.jce.ECDHN");

        config.put("dh", "com.jcraft.jsch.jce.DH");
        config.put("3des-cbc", "com.jcraft.jsch.jce.TripleDESCBC");
        config.put("blowfish-cbc", "com.jcraft.jsch.jce.BlowfishCBC");
        config.put("hmac-sha1", "com.jcraft.jsch.jce.HMACSHA1");
        config.put("hmac-sha1-96", "com.jcraft.jsch.jce.HMACSHA196");
        config.put("hmac-sha2-256", "com.jcraft.jsch.jce.HMACSHA256");
        // The "hmac-sha2-512" will require the key-length 2048 for DH,
        // but Sun's JCE has not allowed to use such a long key.
        //config.put("hmac-sha2-512",  "com.jcraft.jsch.jce.HMACSHA512");
        config.put("hmac-md5", "com.jcraft.jsch.jce.HMACMD5");
        config.put("hmac-md5-96", "com.jcraft.jsch.jce.HMACMD596");
        config.put("sha-1", "com.jcraft.jsch.jce.SHA1");
        config.put("sha-256", "com.jcraft.jsch.jce.SHA256");
        config.put("sha-384", "com.jcraft.jsch.jce.SHA384");
        config.put("sha-512", "com.jcraft.jsch.jce.SHA512");
        config.put("md5", "com.jcraft.jsch.jce.MD5");
        config.put("signature.dss", "com.jcraft.jsch.jce.SignatureDSA");
        config.put("signature.rsa", "com.jcraft.jsch.jce.SignatureRSA");
        config.put("signature.ecdsa", "com.jcraft.jsch.jce.SignatureECDSA");
        config.put("keypairgen.dsa", "com.jcraft.jsch.jce.KeyPairGenDSA");
        config.put("keypairgen.rsa", "com.jcraft.jsch.jce.KeyPairGenRSA");
        config.put("keypairgen.ecdsa", "com.jcraft.jsch.jce.KeyPairGenECDSA");
        config.put("random", "com.jcraft.jsch.jce.Random");

        config.put("none", "com.jcraft.jsch.CipherNone");

        config.put("aes128-cbc", "com.jcraft.jsch.jce.AES128CBC");
        config.put("aes192-cbc", "com.jcraft.jsch.jce.AES192CBC");
        config.put("aes256-cbc", "com.jcraft.jsch.jce.AES256CBC");

        config.put("aes128-ctr", "com.jcraft.jsch.jce.AES128CTR");
        config.put("aes192-ctr", "com.jcraft.jsch.jce.AES192CTR");
        config.put("aes256-ctr", "com.jcraft.jsch.jce.AES256CTR");
        config.put("3des-ctr", "com.jcraft.jsch.jce.TripleDESCTR");
        config.put("arcfour", "com.jcraft.jsch.jce.ARCFOUR");
        config.put("arcfour128", "com.jcraft.jsch.jce.ARCFOUR128");
        config.put("arcfour256", "com.jcraft.jsch.jce.ARCFOUR256");

        config.put("userauth.none", "com.jcraft.jsch.UserAuthNone");
        config.put("userauth.password", "com.jcraft.jsch.UserAuthPassword");
        config.put("userauth.keyboard-interactive", "com.jcraft.jsch.UserAuthKeyboardInteractive");
        config.put("userauth.publickey", "com.jcraft.jsch.UserAuthPublicKey");
        config.put("userauth.gssapi-with-mic", "com.jcraft.jsch.UserAuthGSSAPIWithMIC");
        config.put("gssapi-with-mic.krb5", "com.jcraft.jsch.jgss.GSSContextKrb5");

        config.put("zlib", "com.jcraft.jsch.jcraft.Compressing");
        config.put("zlib@openssh.com", "com.jcraft.jsch.jcraft.Compressing");

        config.put("pbkdf", "com.jcraft.jsch.jce.PBKDF");

        config.put("StrictHostKeyChecking", "ask");
        config.put("HashKnownHosts", "no");

        config.put("PreferredAuthentications", "gssapi-with-mic,publickey,keyboard-interactive,password");

        config.put("CheckCiphers", "aes256-ctr,aes192-ctr,aes128-ctr,aes256-cbc,aes192-cbc,aes128-cbc,3des-ctr,arcfour,arcfour128,arcfour256");
        config.put("CheckKexes", "diffie-hellman-group14-sha1,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521");
        config.put("CheckSignatures", "ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521");

        config.put("MaxAuthTries", "6");
        config.put("ClearAllForwardings", "no");
    }

    private final Vector<Session> sessionPool = new Vector<>();
    private final IdentityRepository defaultIdentityRepository = new LocalIdentityRepository(this);
    private IdentityRepository identityRepository = defaultIdentityRepository;
    private ConfigRepository configRepository = null;
    private HostKeyRepository known_hosts = null;

    public JSch() {
    }

    /**
     * Returns the config value for the specified key.
     *
     * @param key key for the configuration.
     * @return config value
     */
    public static String getConfig(String key) {
        synchronized (config) {
            return (config.get(key));
        }
    }

    /**
     * Sets or Overrides the configuration.
     *
     * @param newconf configurations
     */
    public static void setConfig(java.util.Hashtable<String, String> newconf) {
        synchronized (config) {
            for (java.util.Enumeration<String> e = newconf.keys(); e.hasMoreElements(); ) {
                String key = (e.nextElement());
                config.put(key, (newconf.get(key)));
            }
        }
    }

    /**
     * Sets or Overrides the configuration.
     *
     * @param key   key for the configuration
     * @param value value for the configuration
     */
    public static void setConfig(String key, String value) {
        config.put(key, value);
    }

    public synchronized IdentityRepository getIdentityRepository() {
        return this.identityRepository;
    }

    /**
     * Sets the <code>identityRepository</code>, which will be referred
     * in the public key authentication.
     *
     * @param identityRepository if <code>null</code> is given,
     *                           the default repository, which usually refers to ~/.ssh/, will be used.
     * @see #getIdentityRepository()
     */
    public synchronized void setIdentityRepository(IdentityRepository identityRepository) {
        if (identityRepository == null) {
            this.identityRepository = defaultIdentityRepository;
        } else {
            this.identityRepository = identityRepository;
        }
    }

    public ConfigRepository getConfigRepository() {
        return this.configRepository;
    }

    public void setConfigRepository(ConfigRepository configRepository) {
        this.configRepository = configRepository;
    }

    /**
     * Instantiates the <code>Session</code> object with
     * <code>host</code>.  The user name and port number will be retrieved from
     * ConfigRepository.  If user name is not given,
     * the system property "user.name" will be referred.
     *
     * @param host hostname
     * @return the instance of <code>Session</code> class.
     * @throws JSchException if <code>username</code> or <code>host</code> are invalid.
     * @see #getSession(String username, String host, int port)
     * @see com.jcraft.jsch.Session
     * @see com.jcraft.jsch.ConfigRepository
     */
    public Session getSession(String host)
            throws JSchException {
        return getSession(null, host, 22);
    }

    /**
     * Instantiates the <code>Session</code> object with
     * <code>username</code> and <code>host</code>.
     * The TCP port 22 will be used in making the connection.
     * Note that the TCP connection must not be established
     * until Session#connect().
     *
     * @param username user name
     * @param host     hostname
     * @return the instance of <code>Session</code> class.
     * @throws JSchException if <code>username</code> or <code>host</code> are invalid.
     * @see #getSession(String username, String host, int port)
     * @see com.jcraft.jsch.Session
     */
    public Session getSession(String username, String host) throws JSchException {
        return getSession(username, host, 22);
    }

    /**
     * Instantiates the <code>Session</code> object with given
     * <code>username</code>, <code>host</code> and <code>port</code>.
     * Note that the TCP connection must not be established
     * until Session#connect().
     *
     * @param username user name
     * @param host     hostname
     * @param port     port number
     * @return the instance of <code>Session</code> class.
     * @throws JSchException if <code>username</code> or <code>host</code> are invalid.
     * @see #getSession(String username, String host, int port)
     * @see com.jcraft.jsch.Session
     */
    public Session getSession(String username, String host, int port) throws JSchException {
        if (host == null) {
            throw new JSchException("host must not be null.");
        }
        return new Session(this, username, host, port);
    }

    protected void addSession(Session session) {
        synchronized (sessionPool) {
            sessionPool.addElement(session);
        }
    }

    protected boolean removeSession(Session session) {
        synchronized (sessionPool) {
            return sessionPool.remove(session);
        }
    }

    /**
     * Sets the instance of <code>KnownHosts</code>, which refers
     * to <code>filename</code>.
     *
     * @param filename filename of known_hosts file.
     * @throws JSchException if the given filename is invalid.
     * @see com.jcraft.jsch.KnownHosts
     */
    public void setKnownHosts(String filename) throws JSchException {
        if (known_hosts == null) {
            known_hosts = new KnownHosts();
        }

        if (known_hosts instanceof KnownHosts) {
            synchronized (known_hosts) {
                ((KnownHosts) known_hosts).setKnownHosts(filename);
            }
        }
    }

    /**
     * Sets the instance of <code>KnownHosts</code> generated with
     * <code>stream</code>.
     *
     * @param stream the instance of InputStream from known_hosts file.
     * @throws JSchException if an I/O error occurs.
     * @see com.jcraft.jsch.KnownHosts
     */
    public void setKnownHosts(InputStream stream) throws JSchException {
        if (known_hosts == null) {
            known_hosts = new KnownHosts();
        }
        if (known_hosts instanceof KnownHosts) {
            synchronized (known_hosts) {
                ((KnownHosts) known_hosts).setKnownHosts(stream);
            }
        }
    }

    /**
     * Returns the current hostkey repository.
     * By the default, this method will the instance of <code>KnownHosts</code>.
     *
     * @return current hostkey repository.
     * @see com.jcraft.jsch.HostKeyRepository
     * @see com.jcraft.jsch.KnownHosts
     */
    public HostKeyRepository getHostKeyRepository() {
        if (known_hosts == null) {
            known_hosts = new KnownHosts();
        }
        return known_hosts;
    }

    /**
     * Sets the hostkey repository.
     *
     * @param hostKeyRepository HostKeyRepository
     * @see com.jcraft.jsch.HostKeyRepository
     * @see com.jcraft.jsch.KnownHosts
     */
    public void setHostKeyRepository(HostKeyRepository hostKeyRepository) {
        known_hosts = hostKeyRepository;
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     *
     * @param prvkey filename of the private key.
     * @throws JSchException if <code>prvkey</code> is invalid.
     * @see #addIdentity(String prvkey, String passphrase)
     */
    public void addIdentity(String prvkey) throws JSchException {
        addIdentity(prvkey, (byte[]) null);
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     * Before registering it into identityRepository,
     * it will be deciphered with <code>passphrase</code>.
     *
     * @param prvkey     filename of the private key.
     * @param passphrase passphrase for <code>prvkey</code>.
     * @throws JSchException if <code>passphrase</code> is not right.
     * @see #addIdentity(String prvkey, byte[] passphrase)
     */
    public void addIdentity(String prvkey, String passphrase) throws JSchException {
        byte[] _passphrase = null;
        if (passphrase != null) {
            _passphrase = Util.str2byte(passphrase);
        }
        addIdentity(prvkey, _passphrase);
        if (_passphrase != null) {
            Util.bzero(_passphrase);
        }
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     * Before registering it into identityRepository,
     * it will be deciphered with <code>passphrase</code>.
     *
     * @param prvkey     filename of the private key.
     * @param passphrase passphrase for <code>prvkey</code>.
     * @throws JSchException if <code>passphrase</code> is not right.
     * @see #addIdentity(String prvkey, String pubkey, byte[] passphrase)
     */
    public void addIdentity(String prvkey, byte[] passphrase) throws JSchException {
        Identity identity = IdentityFile.newInstance(prvkey, null, this);
        addIdentity(identity, passphrase);
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     * Before registering it into identityRepository,
     * it will be deciphered with <code>passphrase</code>.
     *
     * @param prvkey     filename of the private key.
     * @param pubkey     filename of the public key.
     * @param passphrase passphrase for <code>prvkey</code>.
     * @throws JSchException if <code>passphrase</code> is not right.
     */
    public void addIdentity(String prvkey, String pubkey, byte[] passphrase) throws JSchException {
        Identity identity = IdentityFile.newInstance(prvkey, pubkey, this);
        addIdentity(identity, passphrase);
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     * Before registering it into identityRepository,
     * it will be deciphered with <code>passphrase</code>.
     *
     * @param name       name of the identity to be used to
     *                   retrieve it in the identityRepository.
     * @param prvkey     private key in byte array.
     * @param pubkey     public key in byte array.
     * @param passphrase passphrase for <code>prvkey</code>.
     */
    public void addIdentity(String name, byte[] prvkey, byte[] pubkey, byte[] passphrase) throws JSchException {
        Identity identity = IdentityFile.newInstance(name, prvkey, pubkey, this);
        addIdentity(identity, passphrase);
    }

    /**
     * Sets the private key, which will be referred in
     * the public key authentication.
     * Before registering it into identityRepository,
     * it will be deciphered with <code>passphrase</code>.
     *
     * @param identity   private key.
     * @param passphrase passphrase for <code>identity</code>.
     * @throws JSchException if <code>passphrase</code> is not right.
     */
    public void addIdentity(Identity identity, byte[] passphrase) throws JSchException {
        if (passphrase != null) {
            try {
                byte[] goo = new byte[passphrase.length];
                System.arraycopy(passphrase, 0, goo, 0, passphrase.length);
                passphrase = goo;
                identity.setPassphrase(passphrase);
            } finally {
                Util.bzero(passphrase);
            }
        }

        if (identityRepository instanceof LocalIdentityRepository) {
            ((LocalIdentityRepository) identityRepository).add(identity);
        } else if (identity instanceof IdentityFile && !identity.isEncrypted()) {
            identityRepository.add(((IdentityFile) identity).getKeyPair().forSSHAgent());
        } else {
            synchronized (this) {
                if (!(identityRepository instanceof IdentityRepository.Wrapper)) {
                    setIdentityRepository(new IdentityRepository.Wrapper(identityRepository));
                }
            }
            ((IdentityRepository.Wrapper) identityRepository).add(identity);
        }
    }

    /**
     * Removes the identity from identityRepository.
     *
     * @param identity the indentity to be removed.
     */
    public void removeIdentity(Identity identity) {
        identityRepository.remove(identity.getPublicKeyBlob());
    }

    /**
     * Lists names of identities included in the identityRepository.
     *
     * @return names of identities
     */
    public Vector<String> getIdentityNames() {
        Vector<String> foo = new Vector<>();
        Vector<Identity> identities = identityRepository.getIdentities();
        for (int i = 0; i < identities.size(); i++) {
            Identity identity = identities.elementAt(i);
            foo.addElement(identity.getName());
        }
        return foo;
    }

    /**
     * Removes all identities from identityRepository.
     */
    public void removeAllIdentity() {
        identityRepository.removeAll();
    }
}
